File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from math import ceil

def solution(fees, records):
    answer = []
    parkDict = {}
    dt, df, at, af = fees
    
    for v in records:
        t,n,s = v.split(" ")
        h, m = map(int,t.split(":"))
        t = h * 60 + m
        if n not in parkDict: parkDict[n] = [0, t, s]
        else:
            if s == "IN": parkDict[n][1], parkDict[n][2] = t, s
            else:
                parkDict[n][0] += t - parkDict[n][1]
                parkDict[n][2] = s
    
    for n,v in parkDict.items():
        f = df
        if v[-1] == "IN": 
            v[0] += 1439 - v[1]
            parkDict[n][2] = "OUT"
        if v[0] > dt: f += ceil((v[0]-dt) / at) * af
        answer.append([n,f])
    
    answer.sort(key = lambda x: x[0])
    return [f for n,f in answer]

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
    const answer = [];
    const parkingTime = {};
    
    records.forEach(v => {
        let [t, n, s] = v.split(' ');
        let [h, m] = t.split(':');
        t = (h * 1) * 60 + (m * 1);
        if (!parkingTime[n]) parkingTime[n] = 0;
        if (s === 'IN') parkingTime[n] += (1439 - t);
        if (s === 'OUT') parkingTime[n] -= (1439 - t);
    });
    
    for (let [n, t] of Object.entries(parkingTime)) {
        if (t <= fees[0]) t = fees[1];
        else t = Math.ceil((t - fees[0]) / fees[2]) * fees[3] + fees[1]
        answer.push([n, t]);
    }
    
    return answer.sort((a, b) => a[0] - b[0]).map(v => v[1]);
}